File: 2021/day2.py
import unittest


def part_one(test_input: str | None = None) -> int:
    with open(test_input if test_input else "day2_input") as f:
        x_value = 0
        depth = 0
        for line in f:
            command, value_str = line.strip().split()
            value = int(value_str)
            match command:
                case "forward":
                    x_value += value
                case "down":
                    depth += value
                case "up":
                    depth -= value

        return x_value * depth


def part_two(test_input: str | None = None) -> int:
    with open(test_input if test_input else "day2_input") as f:
        x_value = 0
        depth = 0
        aim = 0
        for line in f:
            command, value_str = line.strip().split()
            value = int(value_str)
            match command:
                case "forward":
                    x_value += value
                    depth += aim * value
                case "down":
                    aim += value
                case "up":
                    aim -= value

        return x_value * depth


class Test(unittest.TestCase):
    def test_part_one(self):
        self.assertEqual(part_one("day2_test_input"), 150)

    def test_part_two(self):
        self.assertEqual(part_two("day2_test_input"), 900)


if __name__ == "__main__":
    print(part_one())
    print(part_two())

File: 2021/src/bin/day2.rs
use std::fs;

fn part_one(text: &str) -> u32 {
    let mut depth = 0;
    let mut horizontal = 0;
    for line in text.lines() {
        let commands: Vec<&str> = line.split_whitespace().collect();
        let command = commands[0];
        let value: u32 = commands[1].parse().unwrap();

        match command {
            "forward" => horizontal += value,
            "down" => depth += value,
            "up" => depth -= value,
            _ => (),
        }
    }

    depth * horizontal
}

fn part_two(text: &str) -> u32 {
    let mut depth = 0;
    let mut aim = 0;
    let mut horizontal = 0;
    for line in text.lines() {
        let commands: Vec<&str> = line.split_whitespace().collect();
        let command = commands[0];
        let value: u32 = commands[1].parse().unwrap();

        match command {
            "forward" => {
                horizontal += value;
                depth += aim * value;
            }
            "down" => aim += value,
            "up" => aim -= value,
            _ => (),
        }
    }

    depth * horizontal
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_one() {
        let lines = fs::read_to_string("input/day2_test_input").expect("Can't read file");
        assert_eq!(part_one(&lines), 150);
    }

    #[test]
    fn test_part_two() {
        let lines = fs::read_to_string("input/day2_test_input").expect("Can't read file");
        assert_eq!(part_two(&lines), 900);
    }
}

fn main() {
    let lines = fs::read_to_string("input/day2_input").expect("Can't read file");
    println!("{}", part_one(&lines));
    println!("{}", part_two(&lines));
}
